package/environment.js

```diff
@@ -93,7 +104,10 @@ module.exports = class Environment {
       entry: this.entry.toObject(),
 
       module: {
-        rules: this.loaders.values()
+        strictExportPresence: true,
+        rules: [
+          { oneOf: this.loaders.values() }
+        ]
```

Comment:
What do these options change?

Reply:
Previously, we were adding loaders directly to rules array but that had some weird limitations like say, if I want to use some other loader for processing a svg file and we also by default ship the file loader, which also processes svg files. There was no away to ensure the new svg loader is used for svg files, and not file loader. With this change we are saying, use the first loader that matches the criteria (i.e matches test regex or exclude regex). 

Also, the order of processing loaders is now from top to bottom i.e. fixed.

Reply:
https://webpack.js.org/configuration/module/#rule-oneof

Reply:
> **`Rule.oneOf`**
> An array of Rules from which only the first matching Rule is used when the Rule matches.

What if people are relying on files being processed by multiple loaders?

Reply:
Then whenever compiler encounters different format it will traverse again:  

**For example:**  *Started with a `.js` file -> js file has a css import -> look for a loader for css -> then css file has a image linked -> look for a for .png file -> so on....*

Reply:
What if you have `.js` files (or any type) that you want processed by multiple loaders? I assume that's a thing.

Reply:
Ahh right, then that loader will offload to another loader, for example this is a rule for CS2: `use: `['babel-loader', 'coffee-loader']`

Reply:
Each rule defines it's loaders so, this not totally depend on file extension `(.js or .ts)` but underlying loader definition `use`. Some loaders are more implicit like, `vue-loader` it doesn't need that extra `babel-loader` there (figures out itself)

Reply:
If you feel confident about this change then 👍. I haven't seen it used in the wild and I wouldn't be surprised if it breaks something for someone.

Reply:
Yeah, I had some doubts too but after testing this extensively with `hello` examples we ship and the new config seems to work fine 😄  🤞 

Btw, this issue lead this change: https://github.com/rails/webpacker/issues/918#issuecomment-345074512

Previously, it seemed that webpack was using all loaders regardless of order i.e. if you have two loader that has same test extension it won't use the first one but both, atleast that was the case with svg's. This fixes it by enforcing order and only the first matching loader will be used from the array.

Reply:
I have an issue about test coverage config after this change. The rule condition for `istanbul-instrumenter-loader` is different from the rule condition for 'coffee-loader'. The `istanbul-instrumenter-loader` should exclude test files, but the 'coffee-loader' should include the test file. And with `oneOf` option, the two loaders  have to be in the same 'coffee' rule.

Reply:
Hi, this change breaks all combinations like `.coffee.erb` the coffee-loader doesn't use the erb-loader and the erb-loader doesn't use the coffee loader.

As for :
> Previously, we were adding loaders directly to rules array but that had some weird limitations like say, if I want to use some other loader for processing a svg file and we also by default ship the file loader, which also processes svg files.

I guess the workaround for that would be to overwrite the file-loader rule like we do for SASS, something like :

```
const SASSRule = environment.loaders.get("sass");
const SASSLoader = SASSRule.use.find(el => el.loader === "sass-loader");

// Add resolve-url-loader after SASS
SASSRule.use.splice(SASSRule.use.indexOf(SASSLoader), 0, {
  loader: "resolve-url-loader"
});
```

---

This is how I fixed it if anybody runs into the same problem :
```
Environment.prototype._toWebpackConfig = Environment.prototype.toWebpackConfig;

Environment.prototype.toWebpackConfig = function() {
  const webpackConfig = this._toWebpackConfig();

  if (webpackConfig.module.rules[0].oneOf) {
    webpackConfig.module.rules = webpackConfig.module.rules[0].oneOf;
  }

  return webpackConfig;
}

module.exports = environment;
```

Reply:
> Hi, this change (oneOf) breaks all combinations like .coffee.erb the coffee-loader doesn't use the erb-loader and the erb-loader doesn't use the coffee-loader.

Can you report a new issue with this?

Reply:
Oh didn't really look into issues I was looking at the code to figure out why my config was breaking. Thanks @guilleiguaran!

Reply:
Sorry about this folks. There is a PR that should fix all these edge cases. Will merge and make a release as soon as it’s merged and issues are addressed.